src/main/scala/in/ashwanthkumar/suuchi/router/ReplicationRouter.scala

```diff
@@ -66,36 +69,90 @@ abstract class ReplicationRouter(nrReplicas: Int, self: MemberAddress) extends S
-        var iamfound = false
-        eligibleNodes.foreach {
-          case node if node == self => iamfound = true
-          case destination => forward(serverCall.getMethodDescriptor, headers, incomingRequest, destination)
-        }
-        // we need to push this after the forwarding else we return to client immediately saying we're done
-        if (iamfound) {
-          delegate.onMessage(incomingRequest)
-          forwarded = false // FIXME - we had to unset this here since the actual service method is invoked only on onHalfClose()
-        }
+        doReplication(eligibleNodes, serverCall, headers, incomingRequest, delegate)
       case Nil =>
         log.error("This should never happen. No nodes found to place replica")
         serverCall.close(Status.INTERNAL, headers)
     }
   }
+
+  /**
+   * Implement the actual replication logic assuming that you've the right set of nodes.
+   * Just do it!
+   *
+   * Error handling and other scenarios are handled at [[ReplicationRouter.replicate]]
+   * */
+  def doReplication[ReqT, RespT](eligibleNodes: List[MemberAddress], serverCall: ServerCall[ReqT, RespT], headers: Metadata, incomingRequest: ReqT, delegate: Listener[ReqT]): Unit
+}
+
+class SequentialReplicator(nrReplicas: Int, self: MemberAddress) extends ReplicationRouter(nrReplicas, self) {
+  override def doReplication[ReqT, RespT](eligibleNodes: List[MemberAddress], serverCall: ServerCall[ReqT, RespT], headers: Metadata, incomingRequest: ReqT, delegate: Listener[ReqT]) = {
+    log.debug("Sequentially sending out replication requests to the above set of nodes")
+
+    val hasLocalMember = eligibleNodes.exists(_.equals(self))
+
+    eligibleNodes.filterNot(_.equals(self)).foreach { destination =>
+      forward(serverCall.getMethodDescriptor, headers, incomingRequest, destination)
+    }
+
+    // we need to push this after the forwarding else we return to client immediately saying we're done
+    if(hasLocalMember) {
+      delegate.onMessage(incomingRequest)
+      forwarded = false // FIXME - we had to unset this here since the actual service method is invoked only on onHalfClose()
+    }
+  }
+}
+
+object ParallelReplicator {
+  implicit val PARALLEL_REPLICATION_EXECUTOR = Executors.newFixedThreadPool(3)
+}
+class ParallelReplicator(nrReplicas: Int, self: MemberAddress) extends ReplicationRouter(nrReplicas, self) {
+  import ParallelReplicator._
+  override def doReplication[ReqT, RespT](eligibleNodes: List[MemberAddress], serverCall: ServerCall[ReqT, RespT], headers: Metadata, incomingRequest: ReqT, delegate: Listener[ReqT]): Unit = {
+    log.debug("Sending out replication requests to the above set of nodes in parallel")
+
+    val hasLocalMember = eligibleNodes.exists(_.equals(self))
+
+    val replicationResponses = eligibleNodes.filterNot(_.equals(self)).map { destination =>
+        forwardAsync(serverCall.getMethodDescriptor, headers, incomingRequest, destination)
```

Comment:
@brewkode how are we handling errors when one of the futures fail? In the sequential approach it would bail out of the call and the client would get a failed message.